### PR TITLE
Convert to Typescript and Hardhat

### PR DESCRIPTION
**Pull request by 0xbuidler**
Import repo to the hardhat/typescript/ethers waffle stack.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -2,14 +2,24 @@
...
+
+# python
+*.pyc
+
+# hardhat
+/artifacts
+/cache
+/typechain
+
+solidity/build
```
Comment (lbeder) [inline .gitignore]: We probably don't need `*.pyc` and `solidity/build` here at all

File: .gitignore [hunk 2]
```diff
@@ -2,14 +2,24 @@
...
+.coverage_artifacts
+.coverage_contracts
+coverage
+coverageEnv
+coverage.json
+allFiredEvents
+scTopics
```
Comment (lbeder) [inline .gitignore]: Missing newline at the end of the file (but it's really unimportant 😅 )

File: hardhat.config.ts [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+  solidity: {
+    version: '0.6.12',
+    settings: {
+      optimizer: {
+        enabled: true,
+        runs: 200
+      }
+    }
+  }
+};
```
Comment (lbeder) [inline hardhat.config.ts:24]: We should also port the previously defined `mocha` settings in `truffle-config.js`

File: package.json [hunk 4]
```diff
@@ -8,49 +8,45 @@
...
+    "ethers": "^5.0.0",
+    "hardhat": "^2.0.10",
+    "hardhat-typechain": "^0.3.5",
+    "solhint": "^3.3.2",
+    "solidity-coverage": "^0.7.15",
+    "ts-generator": "^0.1.1",
+    "ts-node": "^9.1.1",
+    "typechain": "^4.0.2",
+    "typescript": "^4.1.5",
+    "web3": "^1.3.4"
```
Comment (lbeder) [inline package.json]: Do we still need `web3.js` in the project (since everything else is basically using `ethers`)?

File: package.json [hunk 5]
```diff
@@ -8,49 +8,45 @@
...
+  },
+  "dependencies": {}
```
Comment (lbeder) [inline package.json:49]: 👍 👍 👍

File: scripts/sample-script.ts [hunk 6]
```diff
@@ -0,0 +1,19 @@
...
+    accounts.map((a) => a.address)
+  );
+}
+
+main()
+  .then(() => process.exit(0))
+  .catch((error) => {
+    console.error(error);
+    process.exit(1);
+  });
```
Comment (lbeder) [inline scripts/sample-script.ts]: Do we really need this script?

File: test/tokenGovernance.ts [hunk 7]
```diff
@@ -0,0 +1,262 @@
+import { ethers, Web3 } from 'hardhat';
```
Comment (lbeder) [inline test/tokenGovernance.ts]: In all truffle/hardhat projects I've seen, the test file naming convention was similar to ours: 'Test.js/ts` for `Test.sol`. Have you seen a different standard where the regular js convention was used instead?

File: test/tokenGovernance.ts [hunk 8]
```diff
@@ -0,0 +1,262 @@
...
+import { BigNumber } from 'ethers';
+
+import { expect } from 'chai';
+
+import { TokenGovernance, TokenGovernance__factory } from '../typechain';
+import { MintableToken, MintableToken__factory } from '../typechain';
+import { SignerWithAddress } from '@nomiclabs/hardhat-ethers/dist/src/signer-with-address';
+
+const ZERO_ADDRESS = '0x0000000000000000000000000000000000000000';
```
Comment (lbeder) [inline test/tokenGovernance.ts]: Why not use  `@openzeppelin/test-helpers`?

**Review by lbeder — COMMENTED** [review]
Good stuff, @sshmaxime! 💪

File: scripts/sample-script.ts (hunk 6 above)
Comment (0xbuidler) [reply scripts/sample-script.ts]: We don't indeed, but I thought it was good to have a base script file - even empty (as a base exemple). Also, there was a script before (`export.js`). I didn't get exactly what it was doing so also maybe unsure if I should re-implement it or not. Let me know what should I do

File: package.json (hunk 4 above)
Comment (0xbuidler) [reply package.json]: True, I removed all dep from web3 (needing it for keccak256 hash but doing so from ethers now)

File: test/tokenGovernance.ts (hunk 7 above)
Comment (0xbuidler) [reply test/tokenGovernance.ts]: To be fair I just tried to match as much as possible what was before my PR so I sticked to the same naming (i.e tokenGovernance.js) and I doubled checked [here](https://github.com/bancorprotocol/contracts-solidity/tree/master/solidity/test) as well and it seems the same. I'm unsure if I misunderstood your comment or no :p

File: test/tokenGovernance.ts (hunk 8 above)
Comment (0xbuidler) [reply test/tokenGovernance.ts]: I just tried to stick with the least dependencies so I wasn't feeling having the package just for this (considering also that hardhat integrate directly with mocha to provide us supported matcher for revert and more). Btw, i just figured out that ethers has constants, including the address zero so I removed the line and used address zero from ethers. In addition test-helpers doesn't have any types. Let me know your thoughts

File: package.json (hunk 4 above)
Comment (lbeder) [reply package.json]: 💪

File: scripts/sample-script.ts (hunk 6 above)
Comment (lbeder) [reply scripts/sample-script.ts]: 👍

File: test/tokenGovernance.ts (hunk 7 above)
Comment (lbeder) [reply test/tokenGovernance.ts]: I meant that the file name should start with a capital `T`, i.e. `TokenGovernance.ts`, according to the current conventions.

File: test/tokenGovernance.ts (hunk 8 above)
Comment (lbeder) [reply test/tokenGovernance.ts]: Perfect 👍

File: test/tokenGovernance.ts (hunk 7 above)
Comment (0xbuidler) [reply test/tokenGovernance.ts]: Oh ! my bad ! so yes I misunderstood your comment haha

Comment (lbeder) [reply test/tokenGovernance.ts]: 😅

Comment (0xbuidler) [reply test/tokenGovernance.ts]: Done ✅

**Review by lbeder — APPROVED** [review]

**Pull request merged**